### PR TITLE
Improve error messages for legacy and compressed formats

### LINKED ISSUE #2682: Detect `tszip`ed files and give a helpful error.
Currently opening a tszip file in tskit results in:
`File not in kastore format.`

It's easy to detect these files from their zip `PK` header bytes. We should do this and direct the user to `tszip`.

### PR DESCRIPTION
**Pull request by benjeffery**
Fixes #2682 

Realised as part of this that we didn't have tests for bad streams so added those.

### CODE REVIEW

File: python/tskit/util.py [hunk 1]
```diff
@@ -690,3 +691,32 @@ def random_nucleotides(length: numbers.Number, *, seed: Union[int, None] = None)
...
+        raise tskit.FileFormatError(
+            "The specified file appears to be in HDF5 format. This file "
+            "may have been generated by msprime < 0.6.0 (June 2018) which "
+            "can no longer be read directly. Please convert to the new "
+            "kastore format using the ``tskit upgrade`` command."
+        )
+    open_file.seek(0)
+    if open_file.read(2) == b"\x50\x4b":
+        raise tskit.FileFormatError(
+            "The specified file appears to be in zip format. Use the tszip "
```
Comment (jeromekelleher) [inline python/tskit/util.py:717]: I guess someone could just be passing in a standard zipfile, so should be a little careful there?

**Review by jeromekelleher — COMMENTED** [review]
Looks great. What does the stack trace look like for the various cases? My main issue with using exceptions for flow control is that you end up hard to interpret stack traces.

Should we use ``raise ... from`` to make things easier to follow?

**Comment by benjeffery** [comment]
> What does the stack trace look like for the various cases?

```python
>>> tskit.load("test")
Traceback (most recent call last):
  File "/home/benj/projects/tskit/python/tskit/trees.py", line 4090, in load
    ts.load(
_tskit.FileFormatError: File not in kastore format.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/benj/projects/tskit/python/tskit/trees.py", line 3292, in load
    return TreeSequence.load(
  File "/home/benj/projects/tskit/python/tskit/trees.py", line 4097, in load
    util.raise_known_file_format_errors(file, e)
  File "/home/benj/projects/tskit/python/tskit/util.py", line 716, in raise_known_file_format_errors
    raise tskit.FileFormatError(
_tskit.FileFormatError: The specified file appears to be in zip format, so may be a compressed tree sequence. Try using the tszip module to decompress this file before loading. `pip install tszip; tsunzip <filename>` or use `tszip.decompress` in Python code.
```

File: python/tskit/util.py [hunk 2]
```diff
@@ -690,3 +691,32 @@ def random_nucleotides(length: numbers.Number, *, seed: Union[int, None] = None)
...
+    """
+    Sniffs the file for pk-zip or hdf header bytes, then raises an exception
+    if these are detected, if not raises the existing exception.
+    """
+    # Check for HDF5 header bytes
+    try:
+        open_file.seek(0)
+    except io.UnsupportedOperation:
+        # If we can't seek, we can't sniff the file.
+        raise existing_exception
```
[the remark this answers is not in the document]
Comment (benjeffery) [reply python/tskit/util.py:707]: Nice, done.

File: python/tskit/util.py (hunk 1 above)
Comment (benjeffery) [reply python/tskit/util.py]: I've couched the statement, now reads `The specified file appears to be in zip format, so may be a compressed tree sequence.`

**Comment by benjeffery** [comment]
@mergifyio rebase

**Review by jeromekelleher — APPROVED** [review]

**Comment by benjeffery** [comment]
@mergifyio rebase

**Pull request merged**